internal/tun/tun.go

```diff
@@ -0,0 +1,202 @@
+package tun
+
+// TODO(ainghazal): this package shares a bunch of code with tlsbio, consider
+// refactoring common parts.
+
+import (
+	"bytes"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/apex/log"
+	"github.com/ooni/minivpn/internal/model"
+	"github.com/ooni/minivpn/internal/networkio"
+	"github.com/ooni/minivpn/internal/session"
+)
+
+// StartTUN initializes and starts the TUN device over the vpn.
+func StartTUN(conn networkio.FramingConn, options *model.Options) *TUN {
+	// create a session
+	sessionManager, err := session.NewManager(log.Log)
+	if err != nil {
+		log.WithError(err).Fatal("tun.StartTUN")
+	}
+
+	// create the TUN that will OWN the connection
+	tunnel := newTUN(log.Log, conn, sessionManager)
+
+	// start all the workers
+	workers := startWorkers(log.Log, sessionManager, tunnel, conn, options)
+	tunnel.whenDone(func() {
+		workers.StartShutdown()
+		workers.WaitWorkersShutdown()
+	})
+
+	// Await for the signal from the session manager to known we're ready to start accepting data.
+	// In practice, this means that we already have a valid TunnelInfo at this point
+	// (i.e., three way handshake has completed, and we have valid keys).
+
+	// TODO(ainghazal): we need to timeout here.
+	<-sessionManager.Ready
+	return tunnel
+}
+
+// TUN allows to use channels to read and write. It also OWNS the underlying connection.
+// TUN implements net.Conn
+type TUN struct {
+	logger model.Logger
+
+	// tunDown moves bytes down to the data channel.
+	tunDown chan []byte
+
+	// tunUp moves bytes up from the data channel.
+	tunUp chan []byte
+
+	// conn is the underlying connection.
+	conn networkio.FramingConn
+
+	// hangup is used to let methods know the connection is closed.
+	hangup chan any
+
+	// ensure idempotency.
+	closeOnce sync.Once
+
+	// network is the underlying network for the passed [networkio.FramingConn].
+	network string
+
+	// used to buffer reads from above.
+	readBuffer *bytes.Buffer
+
+	readDeadline  tunDeadline
+	writeDeadline tunDeadline
+
+	session *session.Manager
+
+	// callback to be executed on shutdown.
+	whenDoneFn func()
+}
+
+// newTUN creates a new TUN.
+// This function TAKES OWNERSHIP of the conn.
+func newTUN(logger model.Logger, conn networkio.FramingConn, session *session.Manager) *TUN {
+	return &TUN{
+		logger:        logger,
+		tunDown:       make(chan []byte),
+		tunUp:         make(chan []byte, 10),
+		conn:          conn,
+		hangup:        make(chan any),
+		closeOnce:     sync.Once{},
+		readBuffer:    &bytes.Buffer{},
+		network:       conn.LocalAddr().Network(),
+		readDeadline:  makeTUNDeadline(),
+		writeDeadline: makeTUNDeadline(),
+		session:       session,
+		whenDoneFn: func() {
```

Comment:
```suggestion
		whenDoneFn: func() {
			// nothing
```

We want to clarify to the reader that the function is explicitly and willingly left empty